File: BFS/BFS/Graph.cs
using System.Collections.Generic;

public class Graph
{
    private List<List<int>> adjacencyList;

    public Graph()
    {
        adjacencyList = new List<List<int>>();
    }

    public void AddVertex(int vertex)
    {
        // Đảm bảo chỉ thêm một danh sách rỗng mới khi đỉnh chưa tồn tại
        while (adjacencyList.Count <= vertex)
        {
            adjacencyList.Add(new List<int>());
        }
    }


    public void AddEdge(int startVertex, int endVertex)
    {
        if (startVertex < adjacencyList.Count && endVertex < adjacencyList.Count)
        {
            adjacencyList[startVertex].Add(endVertex);
            adjacencyList[endVertex].Add(startVertex); // Nếu đồ thị vô hướng
        }
    }

    public List<int> GetVertices()
    {
        List<int> vertices = new List<int>();
        for (int i = 0; i < adjacencyList.Count; i++)
        {
            vertices.Add(i);
        }
        return vertices;
    }


    public List<int> DFS(int startVertex)
    {
        bool[] visited = new bool[adjacencyList.Count];
        List<int> result = new List<int>();
        DFSUtil(startVertex, visited, result);
        return result;
    }

    private void DFSUtil(int vertex, bool[] visited, List<int> result)
    {
        visited[vertex] = true;
        result.Add(vertex);
        foreach (var neighbor in adjacencyList[vertex])
        {
            if (!visited[neighbor])
            {
                DFSUtil(neighbor, visited, result);
            }
        }
    }

    public List<int> BFS(int startVertex)
    {
        bool[] visited = new bool[adjacencyList.Count];
        List<int> result = new List<int>();
        Queue<int> queue = new Queue<int>();
        visited[startVertex] = true;
        queue.Enqueue(startVertex);

        while (queue.Count > 0)
        {
            int vertex = queue.Dequeue();
            result.Add(vertex);

            foreach (var neighbor in adjacencyList[vertex])
            {
                if (!visited[neighbor])
                {
                    visited[neighbor] = true;
                    queue.Enqueue(neighbor);
                }
            }
        }

        return result;
    }
}

File: Solve-Path/Solve-Path/src/main/java/com/example/solvePath/Graph.java
package com.example.solvePath;

import java.util.*;

public class Graph {
    private List<List<Integer>> adjacencyList;

    public Graph() {
        adjacencyList = new ArrayList<>();
    }

    public void addVertex(int id) {
        adjacencyList.add(new ArrayList<>());
    }

    public void addEdge(int start, int end) {
        adjacencyList.get(start).add(end);
        adjacencyList.get(end).add(start); // Đồ thị vô hướng
    }

    public List<Integer> dfs(int start) {
        boolean[] visited = new boolean[adjacencyList.size()];
        List<Integer> result = new ArrayList<>();
        dfsUtil(start, visited, result);
        return result;
    }

    private void dfsUtil(int vertex, boolean[] visited, List<Integer> result) {
        visited[vertex] = true;
        result.add(vertex);

        for (int neighbor : adjacencyList.get(vertex)) {
            if (!visited[neighbor]) {
                dfsUtil(neighbor, visited, result);
            }
        }
    }

    public List<Integer> bfs(int start) {
        boolean[] visited = new boolean[adjacencyList.size()];
        List<Integer> result = new ArrayList<>();
        Queue<Integer> queue = new LinkedList<>();

        visited[start] = true;
        queue.offer(start);

        while (!queue.isEmpty()) {
            int vertex = queue.poll();
            result.add(vertex);

            for (int neighbor : adjacencyList.get(vertex)) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    queue.offer(neighbor);
                }
            }
        }
        return result;
    }

    public List<Integer> getNeighbors(int vertex) {
        return adjacencyList.get(vertex);
    }


}

